enhancements/ingress/power-of-two-random-choices.md

```diff
@@ -0,0 +1,247 @@
+	},
+}
+```
+
+Cluster administrators can enable this feature gate as follows:
+
+```yaml
+apiVersion: config.openshift.io/v1
+kind: FeatureGate
+metadata:
+  name: cluster
+spec:
+  customNoUpgrade:
+    enabled:
+    - PowerOfTwoRandomChoicesBalancing
+    # ...
+    disabled:
+    # ...
+  featureSet: CustomNoUpgrade
+```
+
+As was already the case before this enhancement, Routes may override this policy
+using the `haproxy.router.openshift.io/balance` annotation.  Prior to this
+enhancement, the annotation could be set to any one of the values "roundrobin",
+"leastconn", and "source".  This enhancement adds "random" as an allowed option,
+where "random" specifies the Power of Two Random Choices algorithm.  For
+example:
+
+```yaml
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  annotations:
+    haproxy.router.openshift.io/balance: random
+  name: hello-openshift
+  namespace: hello-openshift
+spec:
+  to:
+    kind: Service
+    name: hello-openshift
+```
+
+#### Validation
+
+The featuregates resource is not strictly validated.
+
+The HAProxy configuration template validates the
+`haproxy.router.openshift.io/balance` annotation value.  If a Route specifies an
+annotation with an invalid value, the IngressController uses the default value.
+
+### User Stories
+
+#### As a cluster administrator, I have an IngressController, it has Routes that update frequently, and I want to traffic to be balanced after a reload.
+
+Using the Power of Two Random Choices may improve balancing in this scenario.
+
+#### As a cluster administrator, I have a large number of IngressController Pod replicas and a Route with occasional traffic spikes, and I want to ensure that traffic spikes are balanced across the Route's backend servers.
+
+Using the Power of Two Random Choices satisfies this use-case.
+
+### Implementation Details
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+* openshift/cluster-ingress-operator
+
+The router configures HAProxy using a configuration template.  The template uses
+environment variables as input parameters.  In particular, the template reads
+the `ROUTER_LOAD_BALANCE_ALGORITHM` environment variable to determine the
+default balancing algorithm.  The enhancement modifies the template to allow the
+value "random" to be specified for `ROUTER_LOAD_BALANCE_ALGORITHM` to specify
+the Power of Two Random Choices algorithm.
+
+If the "PowerOfTwoRandomChoicesBalancing" feature gate is not enabled, the
+ingress operator configures the router Deployment by specifying the "leastconn"
+value for `ROUTER_LOAD_BALANCE_ALGORITHM` to specify the Least Connections
+algorithm, preserving the pre-4.8 behavior.
+
+If the feature gate is enabled, the operator configures the Deployment by
+specifying "random" for `ROUTER_LOAD_BALANCE_ALGORITHM`, changing the default
+algorithm to Power of Two Random Choices.
+
```

Comment:
The router also needs to be changed to allow a `random` value for the `haproxy.router.openshift.io/balance` route annotation, right?

Reply:
Yup!  Thanks!